crates/driver/src/domain/competition/solution/trade.rs

```diff
@@ -23,14 +23,164 @@ pub struct Fulfillment {
+        uniform_sell_price: eth::U256,
+        uniform_buy_price: eth::U256,
+    ) -> Result<Self, InvalidFullfilment> {
+        let protocol_fee = {
+            let surplus_fee = match fee {
+                Fee::Static => eth::U256::default(),
+                Fee::Dynamic(fee) => fee.0,
+            };
+
+            let mut protocol_fee = Default::default();
+            for fee_policy in &order.fee_policies {
+                let fee = match fee_policy {
+                    order::FeePolicy::PriceImprovement {
+                        factor,
+                        max_volume_factor,
+                    } => {
+                        match order.side {
+                            order::Side::Buy => {
+                                // How much `sell_token` we need to sell to buy `executed` amount of
+                                // `buy_token`
+                                let executed_sell_amount = executed
+                                    .0
+                                    .checked_mul(uniform_buy_price)
+                                    .ok_or(InvalidFullfilment)?
+                                    .checked_div(uniform_sell_price)
+                                    .ok_or(InvalidFullfilment)?;
+                                // We have to sell slightly more `sell_token` to capture the
+                                // `surplus_fee`
+                                let executed_sell_amount_with_surplus_fee = executed_sell_amount
+                                    .checked_add(surplus_fee)
+                                    .ok_or(InvalidFullfilment)?;
+                                // What is the maximum amount of `sell_token` we are allowed to
+                                // sell based on limit price?
+                                // Equal to full sell amount for FOK orders, otherwise scalled with
+                                // executed amount for partially fillable orders
+                                let limit_sell_amount = order
+                                    .sell
+                                    .amount
+                                    .0
+                                    .checked_mul(executed.0)
+                                    .ok_or(InvalidFullfilment)?
+                                    .checked_div(order.buy.amount.0)
+                                    .ok_or(InvalidFullfilment)?;
+                                // Take protocol fee from the surplus
+                                // Surplus is the diff between the limit price and executed amount
+                                let surplus = limit_sell_amount
+                                    .checked_sub(executed_sell_amount_with_surplus_fee)
+                                    .unwrap_or(eth::U256::zero());
+                                let price_improvement_fee =
+                                    surplus * (eth::U256::from_f64_lossy(factor * 100.)) / 100;
```

Comment:
Have we decided/documented if this factor is supposed to be interpreted as the % of surplus of the "raw solution" (which we only have access to if solvers use our driver) or the % of remaining surplus after the fee is applied. I'd prefer the latter (since this is what we will observe on chain) but here it seems to be the former.

Let me explain with an example (of a driver not using our logic). The protocol fee is 50%. A user wants to sell at most 120 USDC for 100 DAI. On chain we observe 110USDC for 100 DAI. My assumption is that we now charge the solver 5USDC (50% of the observed surplus).

This suggests the true price the solver found (before any fee) must have been something like: sell 105 USDC for 100 DAI

However, if they now used our driver logic with this route, we would take 50% of the 15USDC raw solution surplus and thus withhold 7.5USDC instead of the desired 5.

Reply:
In this equation the `surplus` is calculated as raw surplus - surplus fee, so it's what you prefer. It's because we use executed_sell_amount_**with_surplus_fee** for `surplus` calculation.

Reply:
Then I'm missing something. We can assume the example is on Gnosis Chain (and ignore any network or otherwise pre-specified surplus fee).

How would a route that sells 105 USDC for 100 DAI be adjusted? It would add 7.5 USDC of surplus fee, which would lead to a net settlement of 112.5USDC for 100 DAI. If observed in the autopilot, this would cause a 50% of 7.5USDC = 3.75USDC bill to be sent to the solver. Thus we overcharged the fee.

Looking at the [e2e example](https://www.notion.so/cownation/Fee-Implementation-End-to-end-f8cfc251c63c416aa07aaca1cebdad1b) it does seem that version 1 was agreed (50% protocol fee means we keep 50% of the surplus before fees). However, the example is lacking detail on how the protocol fee is later on recovered from on chain (where a de-facto 50% protocol fee would translate into taking 100% of the observed surplus).

Reply:
We resolved this right? We take 50% of the surplus after fee (surplus_fee)